engine/api/v2_workflow_run_craft.go

```diff
@@ -0,0 +1,469 @@
+			wref.repoCache[vcsName+"/"+repoName] = *repoDB
+		}
+	}
+	if branch == "" {
+		if embeddedEntity || (projKey == wref.run.ProjectKey && entityVCS.ID == wref.runVcsServer.ID && entityRepo.ID == wref.runRepo.ID) {
+			// Get current git.branch parameters
+			branch = wref.run.WorkflowRef
+		} else {
+			defaultCache, has := wref.repoDefaultBranchCache[entityVCS.Name+"/"+entityRepo.Name]
+			if has {
+				branch = defaultCache
+			} else {
+				// Get default branch
+				tx, err := db.Begin()
+				if err != nil {
+					return "", nil, sdk.WithStack(err)
+				}
+				client, err := repositoriesmanager.AuthorizedClient(ctx, tx, store, projKey, entityVCS.Name)
+				if err != nil {
+					return "", nil, err
+				}
+				b, err := client.Branch(ctx, entityRepo.Name, sdk.VCSBranchFilters{Default: true})
+				if err != nil {
+					return "", nil, err
+				}
+				if err := tx.Commit(); err != nil {
+					return "", nil, sdk.WithStack(err)
+				}
+				branch = b.DisplayID
+				wref.repoDefaultBranchCache[entityVCS.Name+"/"+entityRepo.Name] = branch
+			}
+		}
+	}
+
+	completePath := fmt.Sprintf("%s/%s/%s/%s", projKey, vcsName, repoName, entityName)
+	if branch != "" {
+		completePath += "@" + branch
+	}
+
+	switch entityType {
+	case sdk.EntityTypeAction:
+		if _, has := wref.actionsCache[completePath]; has {
+			return completePath, nil, nil
+		}
+	case sdk.EntityTypeWorkerModel:
+		if _, has := wref.workerModelCache[completePath]; has {
+			return completePath, nil, nil
+		}
+	}
+
+	var entityDB *sdk.Entity
+	var err error
+	if projKey != wref.run.ProjectKey || entityVCS.Name != wref.runVcsServer.Name || entityRepo.Name != wref.runRepo.Name || branch != wref.run.WorkflowRef {
+		entityDB, err = entity.LoadByBranchTypeName(ctx, db, entityRepo.ID, branch, entityType, entityName)
+	} else {
+		entityDB, err = entity.LoadByBranchTypeNameCommit(ctx, db, entityRepo.ID, branch, entityType, entityName, wref.run.WorkflowSha)
+	}
+	if err != nil {
+		if sdk.ErrorIs(err, sdk.ErrNotFound) {
+			return "", &sdk.V2WorkflowRunInfo{WorkflowRunID: wref.run.ID, Level: sdk.WorkflowRunInfoLevelWarning, Message: fmt.Sprintf("obsolete workflow dependency used: %s", name)}, nil
+		}
+		return "", nil, err
+	}
+
+	switch entityType {
+	case sdk.EntityTypeAction:
+		var act sdk.V2Action
+		if err := yaml.Unmarshal([]byte(entityDB.Data), &act); err != nil {
+			return "", nil, err
+		}
+		wref.actionsCache[completePath] = act
+	case sdk.EntityTypeWorkerModel:
+		var wm sdk.V2WorkerModel
+		if err := yaml.Unmarshal([]byte(entityDB.Data), &wm); err != nil {
+			return "", nil, err
+		}
+		wref.workerModelCache[completePath] = wm
+	}
+	return completePath, nil, nil
+}
+
+func (api *API) V2WorkflowRunCraft(ctx context.Context, tick time.Duration) error {
+	ticker := time.NewTicker(tick)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case id := <-api.workflowRunCraftChan:
+			api.GoRoutines.Exec(
+				ctx,
+				"V2WorkflowRunCraftChan-"+id,
```

Comment:
keep the same name for the goroutines for both Channel and DB.
I often check the logs this way.
But you can add a log to tell if it's coming from the channel of the DB